docs/datamaps/index.md

```diff
@@ -0,0 +1,164 @@
+# Data Maps
+
+A registry data map contains data-driven, reloadable objects that can be attached to a registry object.  
+This system allows more easily data-driving game behaviour, as they provide functionality such as syncing or conflict resolution, leading to a better and more configurable user experience.  
+
+You can think of tags as entry->boolean maps, while data maps are more flexible entry->object maps.
+
+A data map can be attached to both static, built-in, registries and dynamic data-driven datapack registries.
+
+## Registration
+A data map type should be statically created and then registered to the `RegisterDataMapTypesEvent` (which is fired on the mod event bus). The `DataMapType` can be created using a `DataMapType$Builder`, through `DataMapType#builder`.  
+
+A simple `DataMapType` has two generic arguments: `T` (the values that are being attached) and `R` (the type of the registry the data map is for). A data map of `SomeObject`s that are attached to `Item`s can, as such, be represented as `DataMapType<SomeObject, Item>`.  
+
+Data maps are serialized and deserialized using [Codecs](../datastorage/codecs.md).
+
+Let's take the following record representing the data map value as an example:
+```java
+public record DropHealing(
+        float amount, float chance
+) {
+    public static final Codec<DropHealing> CODEC = RecordCodecBuilder.create(in -> in.group(
+            Codec.FLOAT.fieldOf("amount").forGetter(DropHealing::amount),
+            Codec.floatRange(0, 1).fieldOf("chance").forGetter(DropHealing::chance)
+    ).apply(in, DropHealing::new));
+}
+```
+
+:::warning
+The value should be an *immutable* object, as otherwise weird behaviour can be caused if the object is attached to all values within a tag (since no copy is created).
+:::
+
+For the purposes of this example, we will use this data map to heal players when they drop an item.  
+The `DataMapType` can be created as such:
+```java
+public static final DataMapType<DropHealing, Item> DROP_HEALING = DataMapType.builder(
+        new ResourceLocation("mymod:drop_healing"), Registries.ITEM, DropHealing.CODEC
+).build();
+```
+and then registered to the `RegisterDataMapTypesEvent` using `RegisterDataMapTypesEvent#register`.
+
+## Syncing
+A synced data map will have its values synced to clients. A data map can be marked as synced using `DataMapType$Builder#synced(Codec<T> networkCodec, boolean mandatory)`.   
+The values of the data map will then be synced using the `networkCodec`.  
+If the `mandatory` flag is set to `true`, clients that do not support the data map (including Vanilla clients) will not be able to connect to the server, nor vice-versa. A non-mandatory data map is, on the other side, optional, so it will not prevent any clients from joining.
```

Comment:
They aren't versioned, supporting simply means it is registered on the otherwise and has a non-null network codec.

Reply:
Gotcha. Is there a way to prevent data maps from connecting if they differ in some capacity? Or is this more of a use a network if you need to version or just let it fail?

Reply:
There isn't, no, a dummy channel/versioned payload would be used for that. Optional/mandatory sync is just the bare-bones.

Reply:
Cool.